openlibrary/components/MergeUI/utils.js

```diff
@@ -0,0 +1,156 @@
+    const sources = {};
+    const subsources = {}; // for array elements
+
+    for (let field in result) {
+        sources[field] = [master.key];
+        if (result[field] instanceof Array) {
+            for (let el of result[field]) {
+                subsources[field] = {
+                    [hash_subel(field, el)]: [master.key]
+                };
+            }
+        }
+    }
+
+    for (let dupe of dupes) {
+        for (let field in dupe) {
+            if (!(field in result) && field != 'subtitle') {
+                result[field] = dupe[field];
+                sources[field] = [dupe.key];
+            } else if (result[field] instanceof Array) {
+                result[field] = result[field].concat(dupe[field])
+                sources[field].push(dupe.key);
+            }
+        }
+    }
+
+    // dedup
+    for (let key in result) {
+        if (!(result[key] instanceof Array))
+            continue;
+        switch (key) {
+        case 'authors':
+            const authors = _.cloneDeep(result.authors);
+            authors
+                .filter(a => typeof a.type == 'string')
+                .forEach(a => a.type = { key: a.type });
+            result.authors = _.uniqWith(authors, _.isEqual);
+            break;
+        case 'covers':
+        case 'subjects':
+        case 'subject_people':
+        case 'subject_places':
+        case 'subject_times':
+        case 'excerpts':
+        default:
+            result[key] = _.uniqWith(result[key], _.isEqual);
+            break;
+        }
+    }
+
+    return [result, sources];
+}
+
+export async function do_merge(merged_record, dupes, editions) {
+    editions.forEach(ed => ed.works = [{key: merged_record.key}]);
+    const edits = [
+        merged_record,
+        ...dupes.map(dupe => make_redirect(merged_record.key, dupe)),
+        ...editions
+    ];
+
+    for (let edit of edits) {
+        const r = await post_save(edit, 'merge works');
+        console.log(`completed ${edit.key}; ${r.status}`)
+    }
+}
+
+export function make_redirect(master_key, dupe) {
+    return {
+        location: master_key,
+        key: dupe.key,
+        type: { key: '/type/redirect' }
+    };
+}
+
+export function get_editions(work_key) {
+    const url = `https://openlibrary.org${work_key}/editions.json`;
+    return fetch(url).then(r => r.json());
+}
+
+export function get_lists(key) {
+    const url = `https://openlibrary.org${key}/lists.json`;
+    return fetch(url).then(r => r.json());
+}
+
+export function get_bookshelves(key) {
+    const url = `https://openlibrary.org${key}/bookshelves.json`;
+    return fetch(url).then(r => r.json());
+}
+
+/**
+ * @param {Object} record
+ * @param {string} comment
+ */
+function post_save(record, comment) {
+    const body = new URLSearchParams({
+        body: JSON.stringify(record),
+        _comment: comment,
+        _save: '',
+    });
+    return fetch(`https://openlibrary.org${record.key}.yml?m=edit`, {
```

Comment:
Checking if we want this url hard-coded (assuming this may be addressed in a later commit)

Reply:
It was hand-coded to avoid making edits on testing, where data can be outdated. We should change in the future, but opting to leave as is until our librarians migrate to prod to avoid bad edits!

Reply:
Oh and it looks like I did switch to another API :P `/api/save_many`